File: snippets/fsharp/System/Double/CompareTo/doublesample.fs
module doublesample

open System

//<snippet1>
let d = 4.55
//</snippet1>

//<snippet2>
printfn $"A double is of type {d.GetType()}."
//</snippet2>

//<snippet3>
let mutable completed = false
while not completed do
    printf "Enter a real number: "
    let inp = stdin.ReadLine()
    try
        let d = Double.Parse inp
        printfn $"You entered {d}."
        completed <- true
    with
    | :? FormatException ->
        printfn "You did not enter a number."
    | :? ArgumentNullException ->
        printfn "You did not supply any input."
    | :? OverflowException ->
         printfn $"The value you entered, {inp}, is out of range."
//</snippet3>

//<snippet4>
if d > Double.MaxValue then
   printfn $"Your number is bigger than a double."
//</snippet4>

//<snippet5>
if d < Double.MinValue then
    printfn "Your number is smaller than a double."
//</snippet5>

//<snippet6>
printfn $"Epsilon, or the permittivity of a vacuum, has value {Double.Epsilon}"
//</snippet6>

//<snippet7>
let zero = 0.

// This condition will return false.
if 0. / zero = Double.NaN then
    printfn "0 / 0 can be tested with Double.NaN."
else
    printfn "0 / 0 cannot be tested with Double.NaN use Double.IsNan() instead."
//</snippet7>

//<snippet8>
// This will return true.
if Double.IsNaN(0. / zero) then
    printfn "Double.IsNan() can determine whether a value is not-a-number."
//</snippet8>

//<snippet9>
// This will equal Infinity.
printfn $"10.0 minus NegativeInfinity equals {10.0 - Double.NegativeInfinity}."
//</snippet9>

//<snippet10>
// This will equal Infinity.
printfn $"PositiveInfinity plus 10.0 equals {Double.PositiveInfinity + 10.}."
//</snippet10>

//<snippet11>
// This will return "true".
printfn $"IsInfinity(3.0 / 0) = %b{Double.IsInfinity(3. / 0.)}."
//</snippet11>

//<snippet12>
// This will return "true".
printfn $"IsPositiveInfinity(4.0 / 0) = %b{Double.IsPositiveInfinity(4. / 0.)}."
//</snippet12>

//<snippet13>
// This will return "true".
printfn $"IsNegativeInfinity(-5.0 / 0) = {Double.IsNegativeInfinity(-5. / 0.)}."
//</snippet13>


let a = 500.
//<snippet16>
let obj1 = double 450 |> box
    
if a.CompareTo obj1 < 0 then
    printfn $"{a} is less than {obj1}."

if a.CompareTo obj1 > 0 then
    printfn $"{a} is greater than {obj1}."

if a.CompareTo obj1 = 0 then
    printfn $"{a} equals {obj1}."
//</snippet16>

File: snippets/csharp/System/Double/CompareTo/doublesample.cs
using System;

class DoubleSample
{
    public static void Main()
    {
        //<snippet1>
        double d = 4.55;
        //</snippet1>

        //<snippet2>
        Console.WriteLine($"A double is of type {d.GetType().ToString()}.");
        //</snippet2>

        //<snippet3>
        bool done = false;
        string inp;
        do
        {
            Console.Write("Enter a real number: ");
            inp = Console.ReadLine();
            try
            {
                d = double.Parse(inp);
                Console.WriteLine($"You entered {d.ToString()}.");
                done = true;
            }
            catch (FormatException)
            {
                Console.WriteLine("You did not enter a number.");
            }
            catch (ArgumentNullException)
            {
                Console.WriteLine("You did not supply any input.");
            }
            catch (OverflowException)
            {
                Console.WriteLine($"The value you entered, {inp}, is out of range.");
            }
        } while (!done);
        //</snippet3>

        //<snippet4>
        if (d > double.MaxValue)
            Console.WriteLine("Your number is bigger than a double.");
        //</snippet4>

        //<snippet5>
        if (d < double.MinValue)
            Console.WriteLine("Your number is smaller than a double.");
        //</snippet5>

        //<snippet6>
        Console.WriteLine($"Epsilon, or the permittivity of a vacuum, has value {double.Epsilon.ToString()}");
        //</snippet6>

        //<snippet7>
        double zero = 0;

        // This condition will return false.
        if ((0 / zero) == double.NaN)
            Console.WriteLine("0 / 0 can be tested with Double.NaN.");
        else
            Console.WriteLine("0 / 0 cannot be tested with Double.NaN; use Double.IsNan() instead.");
        //</snippet7>

        //<snippet8>
        // This will return true.
        if (double.IsNaN(0 / zero))
            Console.WriteLine("Double.IsNan() can determine whether a value is not-a-number.");
        //</snippet8>

        //<snippet9>
        // This will equal Infinity.
        Console.WriteLine($"10.0 minus NegativeInfinity equals {(10.0 - double.NegativeInfinity).ToString()}.");
        //</snippet9>

        //<snippet10>
        // This will equal Infinity.
        Console.WriteLine($"PositiveInfinity plus 10.0 equals {(double.PositiveInfinity + 10.0).ToString()}.");
        //</snippet10>

        //<snippet11>
        // This will return "true".
        Console.WriteLine($"IsInfinity(3.0 / 0) == {(double.IsInfinity(3.0 / 0) ? "true" : "false")}.");
        //</snippet11>

        //<snippet12>
        // This will return "true".
        Console.WriteLine($"IsPositiveInfinity(4.0 / 0) == {(double.IsPositiveInfinity(4.0 / 0) ? "true" : "false")}.");
        //</snippet12>

        //<snippet13>
        // This will return "true".
        Console.WriteLine($"IsNegativeInfinity(-5.0 / 0) == {(double.IsNegativeInfinity(-5.0 / 0) ? "true" : "false")}.");
        //</snippet13>

        //<snippet14>
        double a = 500;
        object obj1;
        //</snippet14>

        //<snippet15>
        // The variables point to the same objects.
        object obj2;
        obj1 = a;
        obj2 = obj1;

        if (double.ReferenceEquals(obj1, obj2))
            Console.WriteLine("The variables point to the same double object.");
        else
            Console.WriteLine("The variables point to different double objects.");
        //</snippet15>

        //<snippet16>
        obj1 = (double)450;

        if (a.CompareTo(obj1) < 0)
            Console.WriteLine($"{a.ToString()} is less than {obj1.ToString()}.");

        if (a.CompareTo(obj1) > 0)
            Console.WriteLine($"{a.ToString()} is greater than {obj1.ToString()}.");

        if (a.CompareTo(obj1) == 0)
            Console.WriteLine($"{a.ToString()} equals {obj1.ToString()}.");
        //</snippet16>

        //<snippet17>
        obj1 = (double)500;
        if (a.Equals(obj1))
            Console.WriteLine("The value type and reference type values are equal.");
        //</snippet17>
    }
}
